interfaces/builtin/pkcs11.go

```diff
@@ -0,0 +1,180 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const pkcs11Summary = `allows use of pkcs11 framework and access to exposed tokens`
+
+const pkcs11BaseDeclarationSlots = `
+  pkcs11:
+    allow-installation: false
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+const pkcs11BaseDeclarationPlugs = `
+  pkcs11:
+    allow-installation: true
+    deny-auto-connection: false
+`
+
+type pkcs11Interface struct{}
+
+// Name of the pkcs11 interface.
+func (iface *pkcs11Interface) Name() string {
+	return "pkcs11"
+}
+
+func (iface *pkcs11Interface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              pkcs11Summary,
+		BaseDeclarationSlots: pkcs11BaseDeclarationSlots,
+		BaseDeclarationPlugs: pkcs11BaseDeclarationPlugs,
+	}
+}
+
+func (iface *pkcs11Interface) String() string {
+	return iface.Name()
+}
+
+const pkcs11PermanentSlotSecComp = `
+# Description: Allow operating as an p11-kit server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+
+# Needed by server upon client connect
+accept
+accept4
+`
+
+func (iface *pkcs11Interface) validateSocketPath(path string) error {
+	if err := apparmor_sandbox.ValidateNoAppArmorRegexp(path); err != nil {
+		return fmt.Errorf("pkcs11 interface socket path is invalid: %v", err)
+	}
+	if ok := cleanSubPath(path); !ok {
+		return fmt.Errorf("pkcs11 interface socket path is not clean: %q", path)
+	}
+	return nil
+}
+
+func (iface *pkcs11Interface) getSocketPath(slot *snap.SlotInfo) (string, error) {
+	socketAttr, isSet := slot.Attrs["pkcs11-socket"]
+	if !isSet {
+		return "", fmt.Errorf(`pkcs11 slot %s has missing "pkcs11-socket" attribute`, slot.Name)
```

Comment:
```suggestion
		return "", fmt.Errorf(`cannot use pkcs11 slot %s without "pkcs11-socket" attribute`, slot.Name)
```

Reply:
Error message updated as per suggestion
and new test case added to test for missing attribute